core/protocol/wolkabout/WolkaboutDataProtocol.cpp

```diff
@@ -0,0 +1,418 @@
+/**
+ * Copyright 2021 WolkAbout Technology s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "core/protocol/wolkabout/WolkaboutDataProtocol.h"
+
+#include "core/Types.h"
+#include "core/model/Feed.h"
+#include "core/model/Reading.h"
+#include "core/protocol/wolkabout/WolkaboutProtocol.h"
+#include "core/utilities/Logger.h"
+#include "core/utilities/json.hpp"
+
+#include <memory>
+#include <string>
+
+using nlohmann::json;
+
+namespace wolkabout
+{
+static void to_json(json& j, const Feed& feed)
+{
+    // Check the feed type first
+    auto typeString = toString(feed.getFeedType());
+    if (typeString.empty())
+        throw std::runtime_error("The FeedType value is not valid.");
+    j = json{
+      {"name", feed.getName()}, {"type", typeString}, {"unitGuid", feed.getUnit()}, {"reference", feed.getReference()}};
+}
+
+static void to_json(json& j, const Attribute& attribute)
+{
+    // Check the data type first
+    auto typeString = toString(attribute.getDataType());
+    if (typeString.empty())
+        throw std::runtime_error("The DataType value is not valid.");
```

Comment:
If we are throwing exceptions then I would throw when creating attribute data type (drop the unknown enum value) and not here when we are about to parse it